src/module_5/module5.md

```diff
@@ -0,0 +1,5880 @@
+
+    Parameters:
+    - metrics_df (pd.DataFrame): A DataFrame containing various metrics including 'execution_date'.
+    - all_predicted_tickers (pd.DataFrame): A DataFrame with predicted tickers, including their
+      'execution_date', 'Market_cap', and performance relative to the S&P 500 ('diff_ch_sp500').
+    - top_n_market_cap (int): The number of top tickers to consider based on market cap (default is 500).
+
+    Returns:
+    - pd.DataFrame: The merged DataFrame containing both the original metrics and the baseline for comparison.
+    """
+
+    # Sorting and ranking the tickers based on market cap
+    all_predicted_tickers = all_predicted_tickers.sort_values(
+        ["execution_date", "Market_cap"], ascending=False
+    )  # = False applies to both cols
+    all_predicted_tickers["rank"] = all_predicted_tickers.groupby(
+        ["execution_date"]
+    ).cumcount()
+    top_tickers = all_predicted_tickers[
+        all_predicted_tickers["rank"] <= top_n_market_cap
+    ]
+
+    # Calculate the baseline
+    baseline = (
+        top_tickers.groupby(["execution_date"])["diff_ch_sp500"].mean().reset_index()
+    )
+    baseline = baseline.rename(columns={"diff_ch_sp500": "diff_ch_sp500_baseline"})
+    baseline["execution_date"] = baseline["execution_date"].astype(str)
+
+    metrics_df = pd.merge(metrics_df, baseline, on="execution_date")
+
+    return metrics_df
+```
+
+
+```python
+test_results_final_tree = merge_against_benchmark(
+    test_results_final_tree, all_predicted_tickers
+)
+test_results_final_tree.head()
+```
+
+
+
+
+<div>
+<style scoped>
+    .dataframe tbody tr th:only-of-type {
+        vertical-align: middle;
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>binary_logloss</th>
+      <th>weighted-return</th>
+      <th>n_trees</th>
+      <th>execution_date</th>
+      <th>diff_ch_sp500_baseline</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>0.759610</td>
+      <td>0.099404</td>
+      <td>39</td>
+      <td>2006-06-30</td>
+      <td>0.049213</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>0.730303</td>
+      <td>0.035528</td>
+      <td>39</td>
+      <td>2006-09-30</td>
+      <td>0.067796</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>0.715078</td>
+      <td>-0.052195</td>
+      <td>39</td>
+      <td>2006-12-30</td>
+      <td>0.068473</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>0.710833</td>
+      <td>-0.067471</td>
+      <td>39</td>
+      <td>2007-03-31</td>
+      <td>0.048029</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>0.711701</td>
+      <td>-0.045395</td>
+      <td>39</td>
+      <td>2007-06-30</td>
+      <td>0.077166</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+(
+    ggplot(
```

Comment:
how are you finding ggplot compared to seaborn or matplotlib? I think I told you this in another PR, but I recommend you to also try altair